core/src/main/java/org/apache/calcite/util/format/postgresql/PostgresqlDateTimeFormatter.java

```diff
@@ -38,90 +45,164 @@ public class PostgresqlDateTimeFormatter {
           "FF5"),
       new NumberFormatPattern(
+          ChronoUnitEnum.THOUSANDTHS_OF_MS,
+          0,
+          999_999,
+          6,
           dt -> String.format(Locale.ROOT, "%06d", dt.get(ChronoField.MICRO_OF_SECOND)),
           "FF6"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "AM", "PM"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "am", "pm"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "A.M.", "P.M."),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "a.m.", "p.m."),
-      new NumberFormatPattern(dt -> {
-        final String formattedYear = String.format(Locale.ROOT, "%0,4d", dt.getYear());
-        if (formattedYear.length() == 4 && formattedYear.charAt(0) == '0') {
-          return "0," + formattedYear.substring(1);
-        } else {
-          return formattedYear;
-        }
-      }, "Y,YYY") {
-        @Override protected String trimLeadingZeros(String value) {
-          return value;
-        }
-      },
-      new NumberFormatPattern(
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "AM", "PM"),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "am", "pm"),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "A.M.", "P.M."),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "a.m.", "p.m."),
+      new YearWithCommasFormatPattern(),
+      new NumberFormatPattern(
+          ChronoUnitEnum.YEARS,
+          0,
+          Integer.MAX_VALUE,
+          4,
           dt -> String.format(Locale.ROOT, "%04d", dt.getYear()),
           "YYYY"),
       new NumberFormatPattern(
+          ChronoUnitEnum.YEARS_ISO_8601,
+          0,
+          Integer.MAX_VALUE,
+          4,
           dt -> Integer.toString(dt.get(IsoFields.WEEK_BASED_YEAR)),
           "IYYY"),
       new NumberFormatPattern(
+          ChronoUnitEnum.YEARS_IN_MILLENIA_ISO_8601,
+          0,
+          Integer.MAX_VALUE,
+          3,
           dt -> {
             final String yearString =
                 String.format(Locale.ROOT, "%03d", dt.get(IsoFields.WEEK_BASED_YEAR));
             return yearString.substring(yearString.length() - 3);
           },
           "IYY"),
       new NumberFormatPattern(
+          ChronoUnitEnum.YEARS_IN_CENTURY_ISO_8601,
+          0,
+          Integer.MAX_VALUE,
+          2,
           dt -> {
             final String yearString =
                 String.format(Locale.ROOT, "%02d", dt.get(IsoFields.WEEK_BASED_YEAR));
             return yearString.substring(yearString.length() - 2);
           },
           "IY"),
       new NumberFormatPattern(
+          ChronoUnitEnum.YEARS_IN_MILLENIA,
+          0,
+          Integer.MAX_VALUE,
```

Comment:
why not 1000?

Reply:
Turns out that PostgreSQL does not limit this. It will accept 4 digit numbers.